qa/tasks/cephfs/test_cephfs_shell.py

```diff
@@ -280,3 +296,32 @@ def test_get_to_console(self):
 #            log.info('ls -a succeeded')
 #        else:
 #            log.info('ls -a failed')
+
+    class TestCD(CephFSTestCase):
```

Comment:
I don't see a reason to have another class just for this test case. Let's merge with the other one.

Reply:
I added a separate class to group tests which would allow us trigger a group at a time. The intention was to be able to run tests for `cd`, `ls`, `du`, `df`, etc. separately/individually. It would be inefficient if we trigger entire suite just to test `cd` or some other functionality of shell.

Reply:
I'm okay with this as long as you name the classes properly. You can also subclass the tests so that they inherit from a `TestCephFSShell` abstract class which has the helper methods.

Reply:
> I'm okay with this as long as you name the classes properly. 

`TestCD` is not good? 

> You can also subclass the tests so that they inherit from a TestCephFSShell abstract class which has the helper methods.

The reason why I chose defining class within a class and not inheritance is because defining a class within a class made more sense. `TestCD` conceptually can be said to contained by `TestCephFSShell`. More importantly, this allows invoking all tests when `qa.tasks.cephfs.test_cephfs_shell.TestCephFSShell` is passed.

Reply:
Replying to IRC message that you can't find `test_cd_with_args and `test_cd_with_no_args` in the teuthology test logs, the reason is (probably) because of my "class with class" design. Our test runner just ignores classes inside classes.

If we go with class with class design I would need to write a patch to get the test runner run tests from a class within class supplied to it and if we go for inheritance this change won't be required.

Reply:
> TestCD is not good?

It's okay.

> The reason why I chose defining class within a class and not inheritance is because defining a class within a class made more sense. TestCD conceptually can be said to be contained by TestCephFSShell. More importantly, this allows invoking all tests when qa.tasks.cephfs.test_cephfs_shell.TestCephFSShell is passed.

Ah, that's why we never saw any of your `cd` tests run. It cannot be a "class within a class". It must be a top-level class otherwise the `test_` methods will not be run.

Reply:
> Replying to IRC message that you can't find test_cd_with_args and test_cd_with_no_args` in the teuthology test logs, the reason is (probably) because of my "class with class" design. Our test runner just ignores classes inside classes.

Yes, you got it!

> If we go with class with class design I would need to write a patch to get the test runner run tests from a class within class supplied to it and if we go for inheritance this change won't be required.


Use inheritance please. We do that plenty in other qa tests.

Reply:
Ok. Going ahead with inheritance.

Reply:
Updated the branch. I am using inheritance now. I've moved rest of the methods to a separate classes.

Patrick [wrote](https://github.com/ceph/ceph/pull/28793#discussion_r303159738):
> I'm okay with this as long as you name the classes properly. You can also subclass the tests so that they inherit from a TestCephFSShell abstract class which has the helper methods.

I am a bit confused/unsure about having a abstract class. Won't defining ` TestCephFSShell` as concrete/non-abstract class would be enough?

Reply:
It's fine as is.